modules/tinymce/src/core/main/ts/api/EditorOptions.ts

```diff
@@ -0,0 +1,277 @@
+}
+
+interface ProcessorError {
+  valid: false;
+  message: string;
+}
+
+type SimpleProcessor<T> = (value: T) => boolean;
+type Processor<T, U> = (value: T) => ProcessorSuccess<U> | ProcessorError;
+
+export interface BuiltInOptionTypeMap {
+  'string': string;
+  'number': number;
+  'boolean': boolean;
+  'string[]': string[];
+  'array': any[];
+  'function': Function;
+  'object': any;
+}
+
+export type BuiltInOptionType = keyof BuiltInOptionTypeMap;
+
+interface BaseOptionSpec {
+  immutable?: boolean;
+  deprecated?: boolean;
+  docsUrl?: string;
+}
+
+export interface BuiltInOptionSpec<K extends BuiltInOptionType> extends BaseOptionSpec {
+  processor: K;
+  default?: BuiltInOptionTypeMap[K];
+}
+
+export interface SimpleOptionSpec<T> extends BaseOptionSpec {
+  processor: SimpleProcessor<T>;
+  default?: T;
+}
+
+export interface OptionSpec<T, U> extends BaseOptionSpec {
+  processor: Processor<T, U>;
+  default?: T;
+}
+
+/**
+ * TinyMCE Options API.
+ *
+ * @class tinymce.editor.Options
+ */
+
+export interface Options {
+  /**
+   * Register a new option that can be used within TinyMCE.
+   *
+   * @method register
+   * @param {String} name Name of an option.
+   * @param {OptionSpec} spec A option spec containing how to valid the option and other optional metadata.
+   */
+  register: {
+    <K extends BuiltInOptionType>(name: string, spec: BuiltInOptionSpec<K>): void;
+    <K extends keyof EditorSettings>(name: K, spec: OptionSpec<EditorSettings[K], EditorOptions[K]> | SimpleOptionSpec<EditorSettings[K]>): void;
+    <T, U>(name: string, spec: OptionSpec<T, U>): void;
+    <T>(name: string, spec: SimpleOptionSpec<T>): void;
+  };
+
+  /**
+   * Check if an option has been registered.
+   *
+   * @method isRegistered
+   * @param {String} name Name of an option.
+   * @return {Boolean} True if the option has already been registered, otherwise false.
+   */
+  isRegistered: (name: string) => boolean;
+
+  /**
+   * Get the value of a registered option.
+   *
+   * @method get
+   * @param {String} name Name of a registered option.
+   * @return {Object} An option value or the registered default if not set.
+   */
+  get: {
+    <K extends keyof EditorOptions>(name: K): EditorOptions[K] | undefined;
+    <T>(name: string): T | undefined;
+  };
+
+  /**
+   * Set the value for a registered option.
+   *
+   * @method set
+   * @param {String} name Name of a registered option.
+   * @return {Boolean} True if the option value was successfully set, otherwise false.
+   */
+  set: {
+    <K extends keyof EditorOptions>(name: K, value: EditorOptions[K]): boolean;
+    <T>(name: string, value: T): boolean;
+  };
+
+  /**
+   * Clear any set value for the specified option so that it'll revert back to using the default value.
+   *
+   * @method unset
+   * @param {String} name Name of a registered option.
+   * @return {Boolean} True if the option value was successfully reset, otherwise false.
+   */
+  unset: (name: string) => boolean;
```

Comment:
I'm wondering if this should return something different if it's reverted to a default vs. cleared entirely?

Reply:
I mean the default is just undefined for everything if not defined, so I think this is fine personally as from an end users perspective it won't make a difference.